src/XMakeBuildEngine/Evaluation/LazyItemEvaluator.LazyItemOperation.cs

```diff
@@ -59,9 +59,24 @@ IList<I> GetReferencedItems(string itemType, ImmutableHashSet<string> globsToIgn
                 }
             }
 
-            public abstract void Apply(ImmutableList<ItemData>.Builder listBuilder, ImmutableHashSet<string> globsToIgnore);
+            public virtual void Apply(ImmutableList<ItemData>.Builder listBuilder, ImmutableHashSet<string> globsToIgnore)
+            {
+                var items = SelectItems(listBuilder, globsToIgnore).ToList();
+                MutateItems(items);
```

Comment:
Hmm, even then, the cache would be per LazyItemList instance. For example, the lists representing `[include]` and `[include;update]` would have 2 different GetItems caches because they are 2 objects. So another item pointing to the first list would not get affected by subsequent operations, since those would be in different lazylist objects.

We do have a problem if we decide to do structural sharing between GetItems results on the same item type.